daemon/api.go

```diff
@@ -1082,16 +1085,37 @@ out:
 		origPath = form.Value["snap-path"][0]
 	}
 
-	info, err := readSnapInfo(tempPath)
-	if err != nil {
-		return InternalError("cannot read snap file: %v", err)
-	}
-	snapName := info.Name()
-
 	st := c.d.overlord.State()
 	st.Lock()
 	defer st.Unlock()
 
+	var snapName string
+	var sideInfo *snap.SideInfo
+
+	if !dangerousOK {
+		si, err := snapasserts.DeriveSideInfo(tempPath, assertstate.DB(st))
+		if err != nil && err == asserts.ErrNotFound {
+			msg := "cannot find signatures with metadata for snap"
+			if origPath != "" {
+				msg = fmt.Sprintf("%s %q", msg, origPath)
+			}
+			return BadRequest(msg)
+		}
+		if err != nil {
+			return BadRequest(err.Error())
+		}
+		snapName = si.RealName
+		sideInfo = si
+	} else {
+		// potentially dangerous but force-dangerous was provided
+		info, err := unsafeReadSnapInfo(tempPath)
```

Comment:
I wondered when reading this if we wanted to attempt to derive and fallback only if it errors. Would that be bad? Is there a case for us to force dangerous even if we have the asserts?  Note things like name/etc might change based on it.